src/index.js

```diff
@@ -1,68 +1,135 @@
 'use strict'
 
-const connect = require('pull-ws/client')
+const connect = require('it-ws/client')
 const mafmt = require('mafmt')
 const withIs = require('class-is')
-const Connection = require('interface-connection').Connection
-
 const toUri = require('multiaddr-to-uri')
-const debug = require('debug')
-const log = debug('libp2p:websockets:dialer')
+const { AbortError } = require('abortable-iterator')
+
+const log = require('debug')('libp2p:websockets')
+const assert = require('assert')
 
 const createListener = require('./listener')
+const toConnection = require('./socket-to-conn')
+const { CODE_CIRCUIT, CODE_P2P } = require('./constants')
 
+/**
+ * @class WebSockets
+ */
 class WebSockets {
-  dial (ma, options, callback) {
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Upgrader} options.upgrader
+   */
+  constructor ({ upgrader }) {
+    assert(upgrader, 'An upgrader must be provided. See https://github.com/libp2p/interface-transport#upgrader.')
+    this._upgrader = upgrader
+  }
+
+  /**
+   * @async
+   * @param {Multiaddr} ma
+   * @param {object} [options]
+   * @param {AbortSignal} [options.signal] Used to abort dial requests
+   * @returns {Connection} An upgraded Connection
+   */
+  async dial (ma, options = {}) {
+    log('dialing %s', ma)
+
+    const socket = await this._connect(ma, options)
+    const maConn = toConnection(socket, { remoteAddr: ma, signal: options.signal })
+    log('new outbound connection %s', maConn.remoteAddr)
+
+    const conn = await this._upgrader.upgradeOutbound(maConn)
+    log('outbound connection %s upgraded', maConn.remoteAddr)
+    return conn
+  }
+
+  /**
+   * @private
+   * @param {Multiaddr} ma
+   * @param {object} [options]
+   * @param {AbortSignal} [options.signal] Used to abort dial requests
+   * @returns {Promise<Socket>} Resolves a TCP Socket
```

Comment:
This is actually an extended duplex iterable isn't it? I wouldn't call it a TCP socket, Websocket would be fine although not entirely accurate.